File: services/ingestion-openmeteo/app/Domain/Weather/WeatherIngestion.php
<?php

namespace App\Domain\Weather;

use App\Application\Weather\DTO\WeatherSnapshotData;
use App\Domain\Shared\Event\RecordsEvents;
use App\Domain\Shared\Event\WeatherSnapshotFetched;

/**
 * Agregat koji predstavlja proces ingestije vremenskih podataka za jednu lokaciju.
 *
 * Njegova uloga je da bude mesto u domenu gde nastaje događaj da je
 * vremenski snapshot uspešno preuzet i spreman za dalje procesiranje.
 */
final class WeatherIngestion
{
    use RecordsEvents;

    /**
     * WeatherIngestion constructor.
     *
     * @param Location $location
     */
    public function __construct(private readonly Location $location) {}

    /**
     * Kreira agregat za zadatu lokaciju.
     *
     * @param Location $location
     * @return self
     */
    public static function forLocation(Location $location): self
    {
        return new self($location);
    }

    /**
     * Beleži događaj da je za lokaciju uspešno preuzet i normalizovan snapshot.
     *
     *  Ovaj metod je namerno u domenu, da event ne nastaje u application sloju,
     *  već u agregatu koji predstavlja domen proces ingestije.
     *
     * @param WeatherSnapshotData $snapshotData
     * @return void
     */
    public function recordFetchedSnapshot(WeatherSnapshotData $snapshotData): void
    {
        $this->recordEvent(new WeatherSnapshotFetched(
            location: $this->location,
            snapshotData: $snapshotData,
        ));
    }
}

File: services/ingestion-weatherapi/internal/domain/weather/weather_ingestion.go
package weather

import (
	sharedEvent "ingestion-weatherapi/internal/domain/shared/event"
)

/*
WeatherIngestion predstavlja agregat procesa ingestije vremenskih podataka
za jednu lokaciju.
*/
type WeatherIngestion struct {
	location Location
	recorder sharedEvent.Recorder
}

/*
NewWeatherIngestion pravi novi agregat za zadatu lokaciju.
*/
func NewWeatherIngestion(location Location) *WeatherIngestion {
	return &WeatherIngestion{
		location: location,
	}
}

/*
RecordFetchedSnapshot beleži domain event da je snapshot uspešno preuzet
i normalizovan za lokaciju.
*/
func (w *WeatherIngestion) RecordFetchedSnapshot(snapshot WeatherSnapshotData) {
	w.recorder.Record(
		NewWeatherSnapshotFetched(w.location, snapshot),
	)
}

/*
ReleaseEvents vraća sve zabeležene domain evente iz agregata.
*/
func (w *WeatherIngestion) ReleaseEvents() []sharedEvent.DomainEvent {
	return w.recorder.Release()
}

/*
Location vraća lokaciju za koju agregat radi ingestiju.
*/
func (w *WeatherIngestion) Location() Location {
	return w.location
}
